Sources/NIOPosix/SelectableEventLoop.swift

```diff
@@ -495,15 +535,20 @@ Further information:
             while true {
                 // TODO: Better locking
                 self._tasksLock.withLock { () -> Void in
+                    if !self._immediateTasks.isEmpty {
```

Comment:
So this worries me a little bit. In particular, it affects some of the guarantees we discussed in #1909. In particular, we now have two issues:

1. Immediate tasks enqueued after scheduled ones may execute before them, even when the scheduled task could have executed or was scheduled for now.
2. Immediate tasks can starve scheduled ones altogether.

I think issue 2 is simply a bug that has to be fixed. Issue 1 is more interesting. We could argue we never promised this ordering, and we’d probably be safe, but we have historically suggested that scheduling tasks for now and calling execute weren’t meaningfully different. 

I think if we fix 2 the blast radius of this change isn’t that bad, but if we don’t then it’s really awful.

Reply:
👍 I can fix 2, e.g. zipping immediate and scheduled tasks to guarantee non-starvation.

I could give a stronger ordering guarantee, by e.g. using the same `id` space `ScheduledTask`s have to `UnderlyingTask`s and sort on it before execution. Is that worth the complexity (and performance) cost? I'm sure it's clear, but the semantics would still be complex, in that tasks that get scheduled for later would need to slot in the stream of `UnderlyingTask`s out of `id` order.

Reply:
Done, with the "zipping" approach. Happy to do the (merge) sorting if we need stricter ordering guarantee.